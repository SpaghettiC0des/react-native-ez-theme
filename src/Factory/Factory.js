// @flow
import * as React from "react";

function Factory<T>(
  themeConfig: T
): {
  EzThemeProvider: React.ComponentType<React.ProviderProps<T>>,
  EzThemeConsumer: React.ComponentType<React.ConsumerProps<*>>
} {
  const { Provider, Consumer } = React.createContext<T>(themeConfig);

  const ThemeProvider = ({
    name,
    children
  }: {
    name: string,
    children: any
  }) => {
    const themeName = name.split(".");

    const selectedTheme =
      themeName.length === 1
        ? themeConfig[themeName[0]]
        : themeConfig[themeName[0]][themeName[1]];

    return <Provider value={selectedTheme}>{children}</Provider>;
  };

  const withTheme = (staticItems: Object = {}): React.Node => Component => {
    const WrappedComponent = props => (
      <Consumer>{theme => <Component {...props} theme={theme} />}</Consumer>
    );

    return Object.assign(WrappedComponent, staticItems);
  };

  return {
    EzThemeProvider: ThemeProvider,
    EzThemeConsumer: Consumer,
    withTheme
  };
}

export default Factory;
