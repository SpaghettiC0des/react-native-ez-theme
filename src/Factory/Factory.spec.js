import * as React from "react";
import { Text, View } from "react-native";
import renderer from "react-test-renderer";
import Factory from "./Factory";

describe("Factory", () => {
  it("should accept 1 parameter", () => {
    expect(Factory.length).toEqual(1);
  });

  it("should return an object containing EzThemeProvider, EzThemeConsumer and withTheme", () => {
    const theme = Factory({});

    expect(Object.keys(theme).indexOf("EzThemeProvider") > -1).toBe(true);
    expect(Object.keys(theme).indexOf("EzThemeConsumer") > -1).toBe(true);
    expect(Object.keys(theme).indexOf("withTheme") > -1).toBe(true);
  });

  it("should provide theme config on the child function of EzThemeProvider", () => {
    const themeConfig = {
      DEFAULT: {
        LIGHT: {
          bgColor: "#eee"
        },
        DARK: {
          bgColor: "#333"
        }
      },
      CUSTOM: {
        LIGHT: {
          bgColor: "#fff"
        },
        DARK: {
          bgColor: "#ff0000"
        }
      }
    };
    const { EzThemeConsumer, EzThemeProvider } = Factory(themeConfig);
    const defaultTheme = renderer
      .create(
        <EzThemeProvider name="DEFAULT.LIGHT">
          <EzThemeConsumer>
            {theme => <View style={{ backgroundColor: theme.bgColor }} />}
          </EzThemeConsumer>
        </EzThemeProvider>
      )
      .toJSON();
    expect(defaultTheme).toMatchSnapshot();

    const customTheme = renderer.create(
      <EzThemeProvider name="CUSTOM.DARK">
        <EzThemeConsumer>
          {theme => <View style={{ backgroundColor: theme.bgColor }} />}
        </EzThemeConsumer>
      </EzThemeProvider>
    );

    expect(customTheme).toMatchSnapshot();
  });
});
