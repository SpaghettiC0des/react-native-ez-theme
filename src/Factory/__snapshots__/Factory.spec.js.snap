// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EzThemeProvider should provide theme config on the child function of EzThemeProvider 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#eee",
    }
  }
/>
`;

exports[`EzThemeProvider should provide theme config on the child function of EzThemeProvider 2`] = `
<View
  style={
    Object {
      "backgroundColor": "#ff0000",
    }
  }
/>
`;

exports[`withTheme HOC should add theme props to the wrapped component 1`] = `
<View
  style={
    Object {
      "backgroundColor": "blue",
    }
  }
/>
`;
